File: kuhn.py
import sys
import random
from cfr import CFRTrainer

class KuhnState:
    ACTION_NAMES = ["pass", "bet"]

    def __init__(self, cards=None, history="", actions="pb", current_player=0):
        self.cards = cards
        if self.cards is None:
            self.cards = random.sample((1, 2, 3), 2)
        self.history = history
        self.actions = actions
        self.current_player = current_player

    def is_terminal(self):
        return self.history in ("bp", "bb", "pp", "pbb", "pbp")

    def utility(self):
        if self.history in ("bp", "pbp"):
            return 1

        payoff = 2 if "b" in self.history else 1
        player_higher = self.cards[self.current_player] > self.cards[1-self.current_player]
        return payoff if player_higher else -payoff

    def handle_action(self, action):
        return KuhnState(self.cards[::], self.history+action, self.actions[::], 1-self.current_player)

    def get_representation(self):
        return f"{self.cards[self.current_player]}{self.history}"

if __name__ == "__main__":
    if len(sys.argv) < 2:
        iterations = 20000
    else:
        iterations = int(sys.argv[1])

    trainer = CFRTrainer(KuhnState)

    print(f"running for {iterations//10} iterations")
    trainer.train(iterations//10)
    print("resetting strategies")
    trainer.reset()
    print(f"running for {iterations} iterations")
    value = trainer.train(iterations, True)
    print(f"average value: {value/iterations}")

    for n in trainer.node_map.values():
        print(n)

File: kuhn.jl
include("cfr.jl")
using .CFR
using Random

struct KuhnState
    cards
    history
    actions
    current_player
end

function kuhn_utility(state)
    if state.history in ["bp", "pbp"]
        return 1
    end

    payoff = 'b' in state.history ? 2 : 1
    player_higher = all(state.cards[state.current_player] .>= state.cards[1:2])
    player_higher ? payoff : -payoff
end

kuhn = Game(
    Dict('p' => "pass", 'b' => "bet"),
    () -> KuhnState(shuffle([1,2,3])[1:2], "", "pb", 1),
    (state) -> state.history in ["bp", "bb", "pp", "pbb", "pbp"],
    kuhn_utility,
    (state, action) -> KuhnState(state.cards, state.history*action, state.actions, [2,1][state.current_player]),
    (state) -> string(state.cards[state.current_player],state.history),
    2
)

t = CFRTrainer(kuhn, Dict())
if length(ARGS) > 0
    iterations = parse(Int, ARGS[1])
else
    iterations = 20000
end
println("running for ", iterations÷10, " iterations")
train!(t, iterations÷10)
println("resetting strategies")
reset_nodes(t)
println("running for ", iterations, " iterations")
@time "time tyaken " total_value = train!(t, iterations, true)
print_nodes(t)
println("average game value: ", total_value/iterations)
